File: 0825. Friends Of Appropriate Ages/Solution.cpp
#include <algorithm>
#include <ranges>
#include <vector>
using namespace std;

class Solution {
public:
    int numFriendRequests(vector<int>& ages)
    {
        int cnt[121] = { 0 };
        for (int age : ages)
            ++cnt[age];

        int ans = 0, age_y = 0, cnt_window = 0;
        for (int age_x = 0; age_x < 121; ++age_x) {
            cnt_window += cnt[age_x];
            if (age_y * 2 <= age_x + 14) {
                cnt_window -= cnt[age_y];
                ++age_y;
            }
            if (cnt_window > 0)
                ans += cnt_window * cnt[age_x] - cnt[age_x];
        }
        return ans;
    }
};

File: 0825. Friends Of Appropriate Ages/Solution.py
from typing import List


class Solution:
    def numFriendRequests(self, ages: List[int]) -> int:
        age_cnt = [0] * 121
        for age in ages:
            age_cnt[age] += 1

        ans, cnt, age_y = 0, 0, 14
        for age_x in range(14, 121):
            cnt += age_cnt[age_x]
            if 2 * age_y <= age_x + 14:
                cnt -= age_cnt[age_y]
                age_y += 1
            if cnt > 0:
                ans += age_cnt[age_x] * cnt - age_cnt[age_x]
        return ans
